### DISCUSSION TITLE
Type error: Property 'role' does not exist on type {...}, implementing role base auth

### DISCUSSION THREAD

**Discussion by dorin-js** [nextauthjs/next-auth#5153] [Help]
Following [role-based-login-strategy](https://next-auth.js.org/tutorials/role-based-login-strategy) article, i ran into an issue:
```js
//components/Auth.tsx:35:31
--
Type error: Property 'role' does not exist on type '{ name?: string; email?: string; image?: string; }'
```
Typescript is yelling on `if (session && session.user.role === 'Admin')` at `.role` property.
Anybody, how do i bypass this issue? I  modified local node modules rules, but in production it still won't allow :(

**Comment by tuppers360** [comment]
I am also struggling with this :-( get the same issue as you. 

The docs were updated a while ago I am wondering if there is an issue with this strategy since some updates?

**Reply by dorin-js** [reply]
No answer on stackoverflow either :(
I'l let you know if i find a solution.

**Reply by tuppers360** [reply]
likewise pal :-)

just wondering if you have to extend the model for the user doing something like: [https://next-auth.js.org/getting-started/typescript#extend-default-interface-properties](url) this would then include the role in the expected properties?

I am very new to programming so not sure this is the best way

**Reply by dorin-js** [reply]
yeah, i tried this just now, seems to work, no yelling anymore. ima try in production also
```typescript
import NextAuth, { DefaultSession } from "next-auth"

type Role = "SimpleUser" | "Moderator" | "Admin"
declare module "next-auth" {
  /**
   * Returned by `useSession`, `getSession` and received as a prop on the `SessionProvider` React Context
   */
  interface Session {
    user: {
      /** The user's postal address. */
      role: Role
    } & DefaultSession["user"]
  }
}
```

**Reply by tuppers360** [reply]
awesome if this works could you link your repo? mine is still screaming at me trying to get the role included in the session

**Reply by dorin-js** [reply]
sure, although it has quite a bit of coding there, let me know if u cant find what u need. btw i use Next with Prisma and GitAuthProvider

**Reply by dorin-js** [reply]
https://github.com/dorin-js/mycode-snippets

**Reply by tuppers360** [reply]
As I say i am new to this but could you define an enum in the prisma file like:
`enum Role {
  USER
  ADMIN
}`
 to assign the above value to the db field then use that type in the def file above?

**Reply by dorin-js** [reply]
I am kinda new aswell into coding, half year, first project in Prisma.
In my prisma i have something like 

```js
enum Role {
  SimpleUser
  Moderator
  Admin
}

model User {
  id            String    @id @default(cuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String?
  role          Role      @default(SimpleUser)
  accounts      Account[]
  sessions      Session[]
}
```
but u can't use the same enum outside as there is slightly different syntax, no comas in prisma.schema

in my tsx i have 
```tsx
enum Role {
  SimpleUser = 'SimpleUser',
  Moderator = 'Moderator',
  Admin = 'Admin',
}
```

**Reply by dorin-js** [reply]
![image](https://user-images.githubusercontent.com/98060574/184553579-2c0c286d-927c-4498-901e-51adcbe5be36.png)
that's how it appears in MySQL

**Reply by dorin-js** [reply]
if u make some changes in prisma.schema don't forget to `npx prisma db push` and `npx prisma migrate` to push changes to db and in local node_modules files, and restart your editor, as there is a weird glitch.

**Reply by tuppers360** [reply]
weird I removed the type Role line all works! Love this programming :-)

Just need to do some testing

Thanks for the help and the convo!

**Comment by sylvaincodes** [comment]
//You need to provide custom types for the auth().sessionClaims object, simply redeclare this interface in the global namespace and provide your own custom keys.

export {};

// Create a type for the roles
export type Roles = "admin" | "user";

declare global {
  interface CustomJwtSessionClaims {
    metadata: {
      role?: Roles;
    };
  }
}
